optuna/integration/wandb.py

```diff
@@ -126,10 +155,33 @@ def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -
         metrics = {name: value for name, value in zip(names, trial.values)}
         attributes = {"direction": [d.name for d in study.directions]}
 
-        wandb.config.update(attributes)
-        wandb.log({**trial.params, **metrics}, step=trial.number)
+        run_id = trial._user_attrs.get("run_id")
 
-    def _initialize_run(self) -> None:
-        """Initializes Weights & Biases run."""
+        if not (self._as_sweeps and run_id):
+            run = wandb.run or self._initialize_run()
+            step = trial.number if wandb.run else None
+            run.log({**trial.params, **metrics}, step=step)
+
+        else:
+            # If user is already logging wandb metrics from inside a trial
+            # reuse that run to log these study metrics.
+            run = self._api.run(self._run_path % run_id)
+
+            run.name = f"trial-{trial._trial_id}/{run.name}"
+            run.summary.update({**trial.params, **metrics})
 
-        wandb.init(**self._wandb_kwargs)
+        run.config.update(attributes)
+
+        if self._as_sweeps and wandb.run:
+            run.finish()
+
+    def _initialize_run(self) -> wandb.sdk.wandb_run.Run:
+        """Initializes Weights & Biases run."""
+        run = wandb.init(tags=self._tags, **self._wandb_kwargs)
+        if not isinstance(run, wandb.sdk.wandb_run.Run):
+            raise OSError(
+                "Cannot create a Run."
+                "Expected wandb.sdk.wandb_run.Run as a return."
+                "Got: %s" % (type(run))
+            )
```

Comment:
I think simple assertion is ok here.
```suggestion
        assert isinstance(run, wandb.sdk.wandb_run.Run)
```

Reply:
In my understanding, [`assert`](https://docs.python.org/3/using/cmdline.html) can be ignored by using `-O` option: `python -O sample.py`, so I prefer to `raise`.